### ISSUE TITLE
Mosdepth Coverage per Contig Plot Design/Axis Labels

### ISSUE THREAD

**Issue by schorlton** [ewels/MultiQC#1568]
### Description of bug

When plotting many mosdepth results by "chromosome", it seems that many samples get mixed together and the plot labels don't make sense. For example, the `NZ_` label on the x-axis is a [Veillonella](https://www.ncbi.nlm.nih.gov/nuccore/nz_ap022322), but there's a line for a Gemella:

![image](https://user-images.githubusercontent.com/10245822/138962546-3f570409-1461-4e5c-8945-eae68dc71f6a.png)

### File that triggers the error
[Gemella_haemolysans.mosdepth.global.dist.txt](https://github.com/ewels/MultiQC/files/7421839/Gemella_haemolysans.mosdepth.global.dist.txt)
[Veillonella_sp._S12025_13.mosdepth.global.dist.txt](https://github.com/ewels/MultiQC/files/7421840/Veillonella_sp._S12025_13.mosdepth.global.dist.txt)

### MultiQC Error log

```console
multiqc --version
multiqc, version 1.11
```
```

**Comment by schorlton** [comment]
> As far as I know, `mosdepth` always creates a `summary` file, no? Can you also provide it?

Couldn't tell you, I actually just generate files is `mosdepth.global.dist.txt` [format](https://github.com/brentp/mosdepth#distribution-output), that I believe are compliant with the specified `mosdepth` format. These are the only files that I feed into MultiQC!

**Comment by fgvieira** [comment]
Just made a pull request that tries to fix your issue. Can you double check if it is working now?

**Comment by schorlton** [comment]
Thanks for your help! When I run MultiQC off your fork, using the demo files linked above, I get this:
![image](https://user-images.githubusercontent.com/10245822/139711612-4ac81499-37c2-413c-9ff2-6417181c907a.png)

...if that's correct, then I don't think it's fixed? For example, why does the blue line (the Gemella line), extend onto the Veillonella sequence accession:
![image](https://user-images.githubusercontent.com/10245822/139712013-d662c9c3-f285-4f21-ac7b-2d0d606b5cd6.png)
What are the numbers along the x-axis, and why does the Veillonella have a sequence accession, but the Gemella have numbers?

It's possible that different samples included in the report will have different contig names, so the line for that sample should only overlap the contigs which were found in the `.global.dist.txt` file for that sample. X-axis labels should reflect contig names in the mosdepth outputs. I'm honestly not sure if a line graph makes sense for this plot, because it's theoretically possible to have the same contig in multiple samples, but a contig between them on the plot which is only found in 1 sample, and then the line would have to transverse this label...example:
Purple and red lines are different samples, which both have `Contig_4` in them, while the other contigs are unique to one sample or the other.
![image](https://user-images.githubusercontent.com/10245822/139713851-63bb3a25-bdde-4f38-a3f5-d10b9bf09d83.png)

Something like a violin plot is probably more appropriate, per-contig?

Thanks again!!

**Comment by fgvieira** [comment]
What you are reporting seem to be plotting issue (maybe it assumes all samples to have the same chromosomes), since the underlying data seems to be correct. The same goes with the contig names, not sure why they are changed:
```
defaultdict(<class 'collections.OrderedDict'>, {
'Veillonella_sp._S12025_13': OrderedDict([('NZ_AP022322.1', 0.0)]),
'Gemella_haemolysans': OrderedDict([('NZ_VXKF01000001.1', 0.0), ('NZ_VXKF01000002.1', 0.0), ('NZ_VXKF01000003.1', 0.0), ('NZ_VXKF01000004.1', 0.0), ('NZ_VXKF01000005.1', 0.0), ('NZ_VXKF01000006.1', 0.0), ('NZ_VXKF01000007.1', 0.0), ('NZ_VXKF01000008.1', 0.0), ('NZ_VXKF01000009.1', 6.619999999999903), ('NZ_VXKF01000010.1', 0.0), ('NZ_VXKF01000011.1', 0.0), ('NZ_VXKF01000012.1', 0.0), ('NZ_VXKF01000013.1', 0.0), ('NZ_VXKF01000014.1', 0.0), ('NZ_VXKF01000015.1', 0.0), ('NZ_VXKF01000016.1', 684.1000000000037), ('NZ_VXKF01000017.1', 1417.4300000000308), ('NZ_VXKF01000018.1', 0.0), ('NZ_VXKF01000019.1', 0.0), ('NZ_VXKF01000020.1', 703.8800000000001)])})
```

Not sure why, but maybe @ewels can help.

**Comment by schorlton** [comment]
Hi @fgvieira , just wanted to check in if you had any updates or needed any help? I see you've been making progress in your branch...any chance we could get that merged soon if it's ready? Thanks!

**Comment by fgvieira** [comment]
I have PR #1570 ready and just waiting for @ewels to merge it.

**Comment by ewels** [comment]
PR #1570 doesn't seem to affect this issue as far as I can see.. 😕

**Comment by ewels** [comment]
Oof, this is a particularly horrible plotting bug - the kind of one that gives me nightmares 😰 

In short, the plotting code was overwriting the x-axis categories for each sample. Then the data values use the category index. Here, the dataset with one category just plotted it at index position one. The other then plotted at index position 1-19.. The dataset with a single category happened to be processed last, so there was only one label (the first position) and the rest were defaulting to numerical indices.

I've changed the code to iteratively _add_ categories that are new for each dataset now (instead of overwriting) and then go through the datasets of the each sample. If the category isn't found, `None` is appended to the data, so that the category indexing is valid. With this fix, the plot looks as we expect:

![image](https://user-images.githubusercontent.com/465550/150980174-3b8fd188-c1ad-4b2f-a381-061e2d5f6650.png)

Not many modules use categories for line graphs, and I guess where they do it is most common for all samples to contain the same categories - hence this bug has gone undetected for so long.

Anyway, this should be fixed in 421ef0035385506af6b277c85452435d16cce066 - this is a change to some fairly core code, so if anyone can find some time to test it and make sure that I haven't broken things, that would be great.. 

Thanks for reporting, and thanks for your detective work @fgvieira!

**Comment by schorlton** [comment]
Thanks at @ewels and @fgvieira! That plot looks like it's fixed!
